docs/datastorage/datafixers.md

```diff
@@ -0,0 +1,107 @@
+
+* Renaming the old `HealF` health tag to `Health` (1.9)
+* Replacing all integer ID's in ItemStacks with registry names (~ 1.9)
+* Renaming the ID's of all entities and tile entities from UpperCamel names like `Blaze` to proper registry names like `minecraft:blaze` (1.11)
+* Splitting entities that were variants of each other to separate ID's, e.g. `Elder` tag on `minecraft:guardian` -> `minecraft:elder_guardian` (1.11)
+* Flattening all meta-variants of an Item into separate ID's, e.g. `minecraft:wool` meta 0 -> `minecraft:white_wool` (1.13)
+
+As you can see, this can be used in a very powerful manner to implement any NBT save format fixes you would like to make, without resorting to hacks such as checking for the old tag every tick and replacing it with a new tag.
+
+Terminology and Interfaces
+--------------------------
+
+### IFixType
+This interface represents keys to identify unique situations where NBT tags should be fixed. Its primary implementation is the enum `FixTypes`.
+Example values: `ITEM_INSTANCE` (an individual ItemStack's NBT tag), `CHUNK` (the entire NBT tag of a Chunk), `BLOCK_ENTITY` (the entire NBT tag of a Tile Entity)
+
+### DataFixer
+This is the main entry point for invoking the fixer system, and it also manages registration of fixes and walkers.
+
+### IFixableData
+This interface represents a single "fix" you would like to make, such as one of the examples listed above.
+Fixes are registered to and only run for one `IFixType`.
+
+The interface contains two methods:
+	* `int getVersion()`: This method returns the current save format version of your mod. When the system sees data that has an older version than the result of this method, it will run the fixer.
+	* `NBTTagCompound fixTagCompound(NBTTagCompound compound)`: Accepts the full NBT tag of the requested `IFixType` (i.e. the entire ItemStack, Chunk, or TileEntity). For ItemStacks, the top-level tag is passed, containing the Item ID, stack size, metadata, "tag" NBT tag, and capability NBT tag. The fixed tag should be returned, at the same hierarchy level as it was passed (i.e. return the top-level ItemStack tag).
+
+### IDataWalker
+This interface is registered to and only run for one `IFixType`, and essentially represents the ability to discover other `IFixType`'s within the tag.
+For example, a `TileEntityChest` is of fix type `BLOCK_ENTITY`, but has 27 stacks of contents which are of fix type `ITEM_INSTANCE`. Therefore, an `IDataWalker` is needed to find those stacks and fix them as well.
+
+It has one method:
+	* `NBTTagCompound process(IDataFixer fixer, NBTTagCompound compound, int version)`: Accepts the full NBT tag of the requested `IFixType`, looks for nested `IFixType`'s, and fixes them recursively using `fixer.process()`. 
+For the example above, the walker would look through `compound` for each ItemStack, and then call `fixer.process(FixTypes.ITEM_INSTANCE, stackNBT)` on it, which will recursively invoke all ITEM_INSTANCE fixers and walkers.
```

Comment:
version parameter is missing in the call. Is it the "data version" or the "current mod version of the save format"?

Reply:
walkers will always run all the time. Fixers are only run if the target object is older than the fixer's declared version, so just by running in the fixer you know that you are outdated.

Reply:
that's good to know. That doesn't really answer my question through.

Reply:
ah, sorry, I misread. it's just a typo, the `version` parameter should just be passed on to the fixer